### PR TITLE
feat(sbb-toast): component implementation

### LINKED ISSUE #1026: story(toast): initial implementation
### Feature Description

We want to provide a toast component.

[Figma Spec](https://www.figma.com/file/9r6xSfNmEfCFxl1yFYedrj/Lyne-Components?node-id=2928%3A459049)

**References**
- https://material.angular.io/components/snack-bar/overview
- https://angular.app.sbb.ch/angular/components/notification-toast/overview

**Design Spec**
- Icon is optional and configurable
- Should have an optional close button/icon
- Should have an optional action, which can either be a button or a link
- The positioning of the toast should be configurable (defaults to bottom center)
- @mcilurzo checks max width
  - Message text should wrap if max width exceeded and icon and action should be positioned at the top
- Animation should use ease-in and ease-out to show/hide toast

**Technical Spec**
- Default timeout should be 3 seconds (configurable)
- Inside the `class LyneToast { static { ... } }` anchor the API to the global scope and also export the API
- Check how can the Toast be loaded (via import or standalone usage)

```ts
interface LyneToastConfiguration {
  /** Message to display. */
  message: string;
  /** Hide the toast after defined milliseconds. */
  timeout?: number = 3000;
  /** Either SVG string or reference to a SVG element. */
  icon?: string | HTMLElement;
  /** Id of <template> to use for the icon. */
  iconTemplate?: string;
  /** Action configuration. */
  action?: 'close' | { label: string; action: () => void; } | { label: string; link: string; };
  /** Where the toast should be displayed vertically. Defaults to 'bottom'. */
  verticalPosition?: 'top' | 'bottom' = 'bottom';
  /** Where the toast should be displayed horizontally. Defaults to 'center'. */
  horizontalPosition?: 'left' | 'center' | 'right' | 'start' | 'end' = 'center';
}

globab const LyneToastApi: {
  open(config: LyneToastConfiguration): LyneToastRef;
}
```

### Definition of Done

- [ ] Component is implemented
- [ ] Tests are implemented
- [ ] E2E Tests are implemented
- [ ] Storybook stories are implemented
- [ ] Navigation via keyboard is tested
- [ ] Screen reader output is tested
- [ ] High-contrast is tested
- [ ] Remaining accessibility is tested
- [ ] UX approved

### PR DESCRIPTION
**Pull request by TomMenga**
## Preflight Checklist

- [x] I have read the [Contributing Guidelines](https://github.com/lyne-design-system/lyne-components/blob/master/CONTRIBUTING.md) for this project.
- [x] I agree to follow the [Code of Conduct](https://github.com/lyne-design-system/lyne-components/blob/master/CODE_OF_CONDUCT.md) that this project adheres to.
- [x] I have searched the [pull request tracker](https://github.com/lyne-design-system/lyne-components/pulls) for a Pull Request (PR) that matches the one I want to submit, without success.

## Issue

This PR Closes #1026 

## Pull request checklist

Please check if your PR fulfills the following requirements:

- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been reviewed and added / updated if needed (for bug fixes / features)

See [Review Guidelines](../REVIEW.md) for more information what is checked during review process.

## Changes

The following components were implemented

* sbb-toast

## Browsers

I tested the build on the following browsers:

- [x] Firefox Desktop
- [x] Chrome Desktop
- [x] Edge Desktop
- [x] Safari Desktop
- [x] Chrome Mobile
- [x] Safari Mobile

## Screen readers

I tested the build on the following browsers:

- [x] JAWS Firefox Desktop
- [x] JAWS Chrome Desktop
- [x] NVDA Firefox Desktop
- [x] NVDA Chrome Desktop
- [x] VoiceOver Safari Desktop
- [x] VoiceOver Chrome Desktop
- [x] VoiceOver Safari Mobile
- [x] Android Accessibility Suite Chrome Mobile

## Pull request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [x] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Other (please describe):

## Does this introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

**Review by kyubisation — COMMENTED** [review]
Nice work so far! 😃

File: src/components/sbb-toast/sbb-toast.custom.d.ts
Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.custom.d.ts:1]: I like it :D
Didn't know you could do that though

File: src/components/sbb-toast/sbb-toast.tsx
Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.tsx:210]: Initially, I thought if we should support toasts in shadow DOM but, in the end, I couldn't find a realistic case where this would be needed. But, since you pointed it out, it doesn't hurt to handle those cases too. I will apply your solution :)

File: src/components/sbb-toast/sbb-toast.tsx
Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.tsx:211]: True. Could a method to check for status be useful for consumers? Since they don't have an "official" way to check if the toast is open or closed.

File: src/components/sbb-toast/sbb-toast.tsx
Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.tsx:210]: Hello @kyubisation, 
unfortunately, the "WeakRef" class has been added in ES2021.
Since we are targeting the ES2017, we need to add a reference to a part of the new lib.
I don't fully understand the consequences of this change, do you see a problem that I don't?

image

File: src/components/sbb-toast/sbb-toast.scss
Comment (jeripeierSBB) [inline src/components/sbb-toast/sbb-toast.scss:70]: Allow multilevel redefinition
```suggestion
  z-index: var(--sbb-toast-z-index, var(--sbb-overlay-z-index));
```

File: src/components/sbb-toast/sbb-toast.scss
Comment (jeripeierSBB) [inline src/components/sbb-toast/sbb-toast.scss:116]: I think the size is not fixed for all screen sizes

File: src/components/sbb-toast/sbb-toast.scss
Comment (jeripeierSBB) [inline src/components/sbb-toast/sbb-toast.scss:124]: Should we really target with ::slotted or not just targeting "span" or whatever which is around?

File: src/components/sbb-toast/sbb-toast.scss
Comment (jeripeierSBB) [inline src/components/sbb-toast/sbb-toast.scss:127]: why positioning absoute?

File: src/components/sbb-toast/sbb-toast.stories.tsx
Comment (jeripeierSBB) [inline src/components/sbb-toast/sbb-toast.stories.tsx:101]: would it not more stable to just ust {...args} ? for the position you could still concatenate it. But I'm also thinking if it is only one propery (position), it's better to also only provide one control for it with all the options in it. or otherwise we should make two props in the component

File: src/components/sbb-toast/sbb-toast.tsx
Comment (jeripeierSBB) [inline src/components/sbb-toast/sbb-toast.tsx:36]: I think it should be documented, that setting to "0" you can opt out of a timeout

File: src/components/sbb-toast/sbb-toast.tsx
Comment (jeripeierSBB) [inline src/components/sbb-toast/sbb-toast.tsx:41]:
```suggestion
  /**
   * The name of the icon, choose from the small icon variants
   * from the ui-icons category from here
   * https://icons.app.sbb.ch.
   */
```

File: src/components/sbb-toast/sbb-toast.tsx
Comment (jeripeierSBB) [inline src/components/sbb-toast/sbb-toast.tsx:249]: not sure if i18nCloseDialog describes it best, but it's for sure not wrong :)

File: src/components/sbb-toast/sbb-toast.tsx
Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.tsx:211]: For now remove the api

File: src/components/sbb-toast/sbb-toast.tsx
Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.tsx:254]: The choice was between i18nCloseDialog and i18nCloseAlert. Or maybe create a new one :/
What do you think?

File: src/components/sbb-toast/sbb-toast.scss
Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.scss:120]: I've tested all the resolutions and the "needed" space is constant. 
Mainly because the padding changes with screen resolution and compensates for the change in size of the button

File: src/components/sbb-toast/sbb-toast.scss
Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.scss:128]: That style is specifically meant for slotted sbb-button.
The alternative would be to check for slotted content type and add a class or a data-attribute to target the container only when a button is contained. Not sure it's worth :/
Or maybe there is another way to do that I do not know?

File: src/components/sbb-toast/sbb-toast.stories.tsx
Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.stories.tsx:101]: I thought to have two different controls on storybook because it seemed more convenient to use them.
On the other hand, a single input is more convenient from a code perspective.

But you might be right, it might be misleading and maybe we should have a single control in storybook too. What do you think?

File: src/components/sbb-toast/sbb-toast.scss
Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.scss:131]: Because the sbb-button doesn't fit in the toast unless it ignores the container spacing. 
It was a suggestion that came out in an initial call with Lukas.
Now that you're pointing it out, maybe it is better to work with negative margins to handle this case.
I'll try to implement it and see which one seems less ugly :D

Comment (jeripeierSBB) [reply src/components/sbb-toast/sbb-toast.scss:133]: Ok I understand, please also check the alert component as there is the same problem. Maybe we can align the implementation :)

File: src/components/sbb-toast/sbb-toast.stories.tsx
Comment (jeripeierSBB) [reply src/components/sbb-toast/sbb-toast.stories.tsx:101]: I think I‘d prefer s single control.

File: src/components/sbb-toast/sbb-toast.scss
Comment (jeripeierSBB) [reply src/components/sbb-toast/sbb-toast.scss:130]: Ok I understand. So you can keep it as it is :-)

File: src/components/sbb-toast/sbb-toast.tsx
Comment (jeripeierSBB) [reply src/components/sbb-toast/sbb-toast.tsx]: I agree :-)

File: src/components/sbb-toast/sbb-toast.scss
Comment (jeripeierSBB) [reply src/components/sbb-toast/sbb-toast.scss:122]: Ok, sounds magic 😅

File: src/components/sbb-toast/sbb-toast.stories.tsx
Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.stories.tsx:101]: Got it. This way we can go back using {...args} too

File: src/components/sbb-toast/sbb-toast.scss
Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.scss:133]: Already checked it :)
In that case, it is cleaner to use negative margins since there's no impediment that we, unfortunately, have on the toast.
The impediment has to do with negative bottom-margin ( if you'd like to deep down on it we can have a call )

File: src/components/sbb-toast/sbb-toast.e2e.ts
Comment (jeripeierSBB) [inline src/components/sbb-toast/sbb-toast.e2e.ts:84]: the sbb-toast-close attribute should be described in docs, not?

File: src/components/sbb-toast/sbb-toast.tsx
Comment (jeripeierSBB) [inline src/components/sbb-toast/sbb-toast.tsx:162]: I think it would be a little bit more safe to do the same approach as in tooltip by reading the composedPath. What do you think?

```
event
      .composedPath()
      .find(
        (e) => e instanceof window.HTMLElement && e.hasAttribute('sbb-toast-close')
      ) as HTMLElement;
```

File: src/components/sbb-toast/sbb-toast.tsx
Comment (jeripeierSBB) [inline src/components/sbb-toast/sbb-toast.tsx:161]: completely nitpicky comment, but so far we used `evt` and `event`. `ev` would be new (not completely, it's in autocomplete as well :P . But I let it up to you :D
```suggestion
  @Listen('click') public async onClick(event: Event): Promise<void> {
```

Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.tsx:161]: I'll change those, no problem :)

File: src/components/sbb-toast/sbb-toast.e2e.ts
Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.e2e.ts:84]: It is described D: (I'll improve it a bit)
image

File: src/components/sbb-toast/sbb-toast.tsx
Comment (jeripeierSBB) [reply src/components/sbb-toast/sbb-toast.tsx]: It would include also shadow dom cases. e.g. in sbb-tooltip we applied "sbb-tooltip-close" also to the internal close button (could be done here as well). the parents should better not be included as you said. Lukas and I had the idea of creating a helper function doing this. something like `composedPathHasAttribute(event, attributeName, endElement)`. This function would filter out all elements after/above endElement and then check for an attribute. And as you mentioned it would also include setting sbb-tooltip-close on in between elements, like on a container div containing several actions.

Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.tsx]: Thanks for the explanation :D
I'll align the implementation to the tooltip

Comment (jeripeierSBB) [reply src/components/sbb-toast/sbb-toast.tsx]: I think we had a little misunderstanding here, can you create a helper function for both the tooltip and toast as following: `composedPathHasAttribute(event, attributeName, endElement)` (or similar). The function should exclude all parents (starting from endElement which would be sbb-tooltip, or sbb-toast). This fixes a rare case issue where parent elements are searched too and helps avoiding code duplication. Thanks!

Comment (TomMenga) [reply src/components/sbb-toast/sbb-toast.tsx]: My bad :( 
I thought that the helper method was a future refactoring. I'll fix this asap

File: src/global/helpers/composed-path-has-attribute.ts
Comment (jeripeierSBB) [inline src/global/helpers/composed-path-has-attribute.ts:13]: Sorry, hopefully last comment, but it can also be null which should be reflected in types

**Review by jeripeierSBB — APPROVED** [review]
LGTM, thanks! Let's wait for Manuels approval

**Review by jeripeierSBB — APPROVED** [review]
Thanks for the improvements! LGTM, waiting for reply of Mürsel

**Pull request merged**
